src/index.ts

```diff
@@ -234,47 +255,75 @@ export async function ingest(
+            );
+
+            const extract = extractor.extract_quads(data);
+            const indexOperations: IndexBulkOperations =
+                repository.prepareIndexBulk();
+
+            await handleRecords(extract, repository, indexOperations);
+            await handleRelations(extract, repository, indexOperations);
+            await handleBuckets(extract, repository, indexOperations);
+
+            await repository.ingestIndexBulk(indexOperations);
+        }
         ingestData = false;
+        logger.info("[Ingest] Data stream closed.");
         return await closeRepository();
-    });
+    };
 
-    metadata.on("end", async () => {
+    // Helper to process one iterator
+    const processMetadataIterator = async (iter: any) => {
+        for await (const meta of iter) {
+            logger.debug(`[Ingest] -- processing metadata entry\n${meta}`)
+            await handleMetadata(meta, repository, ingestMetadata)
+        }
         ingestMetadata = false;
         return await closeRepository();
-    });
-
-    metadata.data(async (meta) => await handleMetadata(meta, repository, ingestMetadata));
-
-    if (metadata.lastElement) {
-        await handleMetadata(metadata.lastElement, repository, ingestMetadata);
-    }
-    logger.debug("Attached metadata handler");
-
-    await repository.createIndices();
+        
+    };
 
-    const extractor = new Extractor();
+    processDataIterator(data.strings())
+    processMetadataIterator(metadata.strings())
+}
 
-    data.data(async (input: RDF.Quad[] | string) => {
-        const data = maybe_parse(input);
-        if (!ingestData) {
-            logger.error("Cannot handle data, repository is closed");
-            return;
-        }
-        logger.debug(
-            `Handling ingest for '${data.find((q) => q.predicate.equals(SDS.terms.payload))?.object?.value}'`,
-        );
 
-        const extract = extractor.extract_quads(data);
-        const indexOperations: IndexBulkOperations =
-            repository.prepareIndexBulk();
 
-        await handleRecords(extract, repository, indexOperations);
-        await handleRelations(extract, repository, indexOperations);
-        await handleBuckets(extract, repository, indexOperations);
 
-        await repository.ingestIndexBulk(indexOperations);
-    });
+type Args = {
+    dataInput: Reader,
+    metadataInput: Reader,
+    database: DBConfig,
+};
 
-    logger.debug("Attached data handler");
+export class Ingest extends Processor<Args> {
+    
+    repository: Repository;
+    async init(this: Args & this): Promise<void> {
+        this.repository = await ingestInit(
+            this.dataInput, 
+            this.metadataInput, 
+            this.database
+        )
+        logger.debug('[Ingest] initialized the Ingest processor.')
+    }
+    async transform(this: Args & this): Promise<void> {
+        ingestTransform(this.dataInput, this.metadataInput, this.repository)
```

Comment:
Let’s move the complete implementation to this class, instead of simply calling the old functions from here.